src/main/java/org/openrefine/extensions/commons/importer/RelatedCategoryFetcher.java

```diff
@@ -0,0 +1,125 @@
+                .addQueryParameter("action", "query")
+                .addQueryParameter("prop", "categories")
+                .addQueryParameter("titles", titles)
+                .addQueryParameter("format", "json").build();
+        Request request = new Request.Builder().url(urlRelatedCategories).build();
+        Response response = client.newCall(request).execute();
+        JsonNode jsonNode = new ObjectMapper().readTree(response.body().string());
+        List<JsonNode> relatedCategories = new ArrayList<>();
+        toCategoriesColumn = new ArrayList<>();
+        for (int i = 0; i < fileRecordOriginal.size(); i++) {
+            relatedCategories.add(jsonNode.path("query").path("pages").path(fileRecordOriginal.get(i).pageId).path("categories"));
+            List<String> categoriesPerFile = new ArrayList<>();
+            for (int j = 0; j < relatedCategories.get(i).size(); j++) {
+                categoriesPerFile.add(relatedCategories.get(i).get(j).findValue("title").asText());
+            }
+            toCategoriesColumn.add(categoriesPerFile);
+        }
+
+        return toCategoriesColumn;
+    }
+
+    /*
+     * Returns {@code true} if the iteration has more elements for
+     * which to fetch related categories.
+     * (In other words, returns {@code true} if {@link #next} would
+     * return an element rather than throwing an exception.)
+     *
+     * @return {@code true} if the iteration has more elements
+     */
+    @Override
+    public boolean hasNext() {
+        return iteratorFileRecords.hasNext();
+    }
+
+    /*
+     * This method iterates over each of the categories related to a file
+     * and stores them as a list in the relatedCategories parameter of
+     * each file record
+     *
+     * @return an instance of the FileRecord updated to include its related categories
+     */
+    @Override
+    public FileRecord next() {
+
+        if (fileRecordNewIndex <= 0) {
+            int filesIndex = 0;
+            if (filesIndex < API_TITLES_LIMIT) {
+                while (iteratorFileRecords.hasNext()) {
+                    fileRecordOriginal.add(iteratorFileRecords.next());
+                    filesIndex++;
+                }
+            }
+            try {
+                getRelatedCategories(fileRecordOriginal);
+            } catch (IOException e) {
+                // FIXME
+                e.printStackTrace();
+            }
+            for (int i = 0; i < fileRecordOriginal.size(); i++) {
+                fileRecordNew.add(new FileRecord(fileRecordOriginal.get(i).fileName, fileRecordOriginal.get(i).pageId,
+                        toCategoriesColumn.get(i), null));
+                fileRecordNewIndex++;
+            }
+            if (iteratorFileRecords.hasNext()) {
+                filesIndex = 0;
+            }
+            return fileRecordNew.get(--fileRecordNewIndex);
+        } else {
+            return fileRecordNew.get(--fileRecordNewIndex);
```

Comment:
The fact that you return the file records by decreasing index suggests that you might not return the file records in their original order (I have not checked). It would be cleaner to only use increasing indices.